File: Java/1074. Number of Submatrices That Sum to Target.java
class Solution {
    public int numSubmatrixSumTarget(int[][] matrix, int target) {
        int R = matrix.length, C = matrix[0].length;
        
        int[][] sums = new int[R+1][C+1];
        for (int r=1; r<R+1; r++) {
            for (int c=1; c<C+1; c++) {
                sums[r][c] = sums[r][c-1] + sums[r-1][c] - sums[r-1][c-1] + matrix[r-1][c-1];
            }
        }
        
        int count = 0, cur;
        Map<Integer, Integer> m = new HashMap();
        for (int r1=1; r1<=R; ++r1) {
            for (int r2=r1; r2<=R; ++r2) {
                m.clear();
                m.put(0, 1);
                
                for (int c=1; c<=C; ++c) {
                    cur = sums[r2][c] - sums[r1-1][c];
                    count += m.getOrDefault(cur-target, 0);
                    m.put(cur, m.getOrDefault(cur, 0)+1);
                }
            }
        }
        
        return count;
    }
}

File: Python3/1074. Number of Submatrices That Sum to Target.py
class Solution:
    def numSubmatrixSumTarget(self, matrix: List[List[int]], target: int) -> int:
        R, C = len(matrix), len(matrix[0])
        
        # compute 2D prefix sums
        ps = [[0] * (C+1) for _ in range(R+1)]
        for r in range(1, R+1):
            for c in range(1, C+1):
                ps[r][c] = ps[r][c-1] + ps[r-1][c] - ps[r-1][c-1] + matrix[r-1][c-1]
                
        count = 0
        # reduce to 1D prefix sum problem
        for r1 in range(1, R+1):
            for r2 in range(r1, R+1):
                sums = defaultdict(int)
                sums[0] = 1
                
                for c in range(1, C+1):
                    # current prefix sum
                    cur = ps[r2][c] - ps[r1-1][c]
                    count += sums[cur - target]
                    
                    # incremetn current prefix sum's count
                    sums[cur] += 1
                    
        return count